x52output.1
.\" x52output manual page
.\" Copyright (C) 2007 Eduard Hasenleithner <eduard@hasenleithner.at>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

.TH x52output 1
.SH NAME
x52output \- command line frontend for the x52pro library
.SH SYNOPSIS
.nf
\f3x52output text 0 'First Line'\f1
\f3x52output text 1 'Second Line'\f1
\f3x52output text 2 'Third Line'\f1
\f3x52output bri 0 40\f1
\f3x52output bri 1 50\f1
\f3x52output led 1 1\f1         # not available for old X52
\f3x52output time 1 12 20\f1
\f3x52output date 7 10 17\f1
\f3x52output second 30\f1       # YOKE only
\f3x52output offset 0 1 0 90\f1 # set 90 minutes offset on time display 2
.fi
.SH DESCRIPTION
.I x52output
is a command line frontend for the x52pro MFD library. It is intended
for debugging and/or scripting purposes.

.PP
The x52pro MFD
library provides access to the display/LEDS of following SAITEK
pro flight devices
.IP \(bu
X52 Pro Flight Control System
.IP \(bu
X52 Flight Control System
.IP \(bu
X52 Pro Flight Yoke.

.PP
.I x52output
has following sub-commands:
.TP
\f3text\f1 \f2line\f1 \f2text\f1
Outputs a single line of text to one of the three lines on the
x52/x52pro MFD. \f2line\f1 is a number from 0 to 2, \f2text\f1
is an alphanumeric text of length 16. Some graphical characters
are also supported but no work has currently been spent on
charset compatibility.

.TP
\f3bri\f1 \f2mfd\f1 \f2brightness\f1
Sets the brightness of either the button LEDs (x52pro only) or the
MFD. \f2mfd\f1=0 selects the buttons and \f2mfd\f1=1 selects the MFD.
\f2brightness\f1 is a decimal value in the range from 0 to 127.

.TP
\f3led\f1 \f2led-idx\f1 \f2color\f1
Sets the led color of one of the x52pro buttons. \f2led-idx\f1 is a
value from 1 to 19. The meaning of the \f2led-idx\f1 values are
specified in the next paragraph.

.TP
\f3time\f1 \f2h24\f1 \f2hour\f1 \f2minute\f1
Sets the time shown on the MFD. \f2h24\f1=0 displays the time in
AM/PM notation, \f2h24\f1=1 displays the time in 24 Hour notation.
\f2hour\f1 is a value from 0 to 24 while \f2minute\f1 is a value
from 0 to 60.

.TP
\f3date\f1 \f2year\f1 \f2month\f1 \f2mday\f1
Sets the date shown on the MFD. \f2year\f1, \f2month\f1, and \f2mday\f1
are just values from 0 to 99. Using this order will show as YY-MM-DD on
the MFD but it is possible to use the three values in any order preferred.

.TP
\f3second\f1 \f2second\f1
Sets the second value shown on the Pro Flight Yoke MFD. Can be a value
from 0 to 59.

.TP
\f3offset\f1 \f2offset\f1
Sets the time offset of the secondary and tertiary time display of the
x52/x52pro MFD. \f2idx\f1=0 selects the secondary display while
\f2idx\f1=1 selects the tertiary display. \f2neg\f1=0 sets a positive offset,
\f2neg\f1=1 sets a negative offset.
\f2offset\f1 is in minutes from 0 to 1023.

.PP
\f2led-idx\f1 for the \f3led\f1 sub-command can be one of the following values:
.TP
.B 1
LED(s) of Fire Button
.TP
.B 2
Red LED of A Button
.TP
.B 3
Green LED of A Button
.TP
.B 4
Red LED of B Button
.TP
.B 5
Green LED of B Button
.TP
.B 6
Red LED of D Button
.TP
.B 7
Green LED of D Button
.TP
.B 8
Red LED of E Button
.TP
.B 9
Green LED of E Button
.TP
.B 10
Red LED of T1 Switch
.TP
.B 11
Green LED of T1 Switch
.TP
.B 12
Red LED of T2 Switch
.TP
.B 13
Green LED of T2 Switch
.TP
.B 14
Red LED of T3 Switch
.TP
.B 15
Green LED of T3 Switch
.TP
.B 16
Red LED of Coolie Hat
.TP
.B 17
Green LED of Coolie Hat
.TP
.B 18
Red LED of i Button
.TP
.B 19
Green LED of i Button

.PP
.I x52output
needs write access to the raw x52 USB device. This package is supplied
with udev rules for providing write access on the SAITEK MFD joysticks
to the plugdev group members. If
.I x52output
fails, permission problems can be checked by means of running it with sudo.

.SH EXIT STATUS
If the joystick/yoke is not found
.I x52output
will return a status
code of 1. In every other situation it will return a status code of 0.
This includes malformed command lines.
.SH BUGS
.I x52output
can currently control only a single SAITEK mfd device. If there are
more than one devices attached to the system, the first will always
be used.
.SH AUTHOR
Written by Eduard Hasenleithner <eduard@hasenleithner.at>


changelog
0.1.2 (2012-07-21)
  Adding GPL license headers to x52output.c and x52output.1
  Adding GPL section to COPYING

0.1.1 (2008-01-22)
  Changed remaining references from x52test to x52output
  Moved manpage for x52output to correct section 1
  Fixed install section of makefile
  Added another device-id (0x0255) for the old x52 Joystick
  Changed udev rules to resemble the "45-libsane.rules" file

0.1.0 (2008-01-21)
  Merged libraryfication makefile from ubuntu.
  Renamed x52test to x52output to generally expose its usability.
  Added manpage for x52output
  Added udev rules for giving plugdev members access.
  Added debug enable function for suppressing debug output.

0.0.3 (2007-12-20)
  Implemented secondary and tertiary time display

0.0.2 (2007-12-17)
  Added support for old X52 and YOKE
  Added Check functions if particular output function is possible for
    the connected HID model.
  Updated README for new "seconds" command

0.0.1 (2007-11-04)
  Initial release


x52output.c
/*
 * x52output utility
 * Copyright (C) 2007 Eduard Hasenleithner <eduard@hasenleithner.at>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <x52pro.h>

/* just for debugging */
int x52_custom(struct x52 *x52, int index, int value);

int main(int argc, char**argv)
{
	struct x52 *hdl = x52_init();
	if (hdl==NULL)
		return 1;
	x52_debug(hdl, 1);

	char *prog = argv[0];

	if (argc < 2) {
		printf("usage: %s {text|bri|led|time|offset|date|custom}\n", prog);
		return 0;
	}
	char *cmd = argv[1];

	if (!strcmp(cmd, "text")) {
		if (argc < 4) {
			printf("usage: %s text <line> <text>\n", prog);
			return 0;
		}
		x52_settext(hdl, atoi(argv[2]), argv[3], strlen(argv[3]));
	} else if (!strcmp(cmd, "bri")) {
		if (argc < 4) {
			printf("usage: %s bri <mfd> <value>\n", prog);
			return 0;
		}
		x52_setbri(hdl, atoi(argv[2]), atoi(argv[3]));
	} else if (!strcmp(cmd, "led")) {
		if (argc < 4) {
			printf("usage: %s led <id> <on>\n", prog);
			return 0;
		}
		x52_setled(hdl, atoi(argv[2]), atoi(argv[3]));
	} else if (!strcmp(cmd, "time")) {
		if (argc < 5) {
			printf("usage: %s time <h24> <hour> <min>\n", prog);
			return 0;
		}
		x52_settime(hdl, atoi(argv[2]), atoi(argv[3]), atoi(argv[4]));
	} else if (!strcmp(cmd, "offset")) {
		if (argc < 6) {
			printf("usage: %s time <idx> <h24> <inv> <minutes>\n", prog);
			return 0;
		}
		x52_setoffs(hdl, atoi(argv[2]), atoi(argv[3]), atoi(argv[4]),
			atoi(argv[5]));
	} else if (!strcmp(cmd, "date")) {
		if (argc < 5) {
			printf("usage: %s date <year> <month> <day>\n", prog);
			return 0;
		}
		x52_setdate(hdl, atoi(argv[2]), atoi(argv[3]), atoi(argv[4]));
	} else if (!strcmp(cmd, "second")) {
		if (argc < 3) {
			printf("usage: %s date <year> <month> <day>\n", prog);
			return 0;
		}
		x52_setsecond(hdl, atoi(argv[2]));
	} else if (!strcmp(cmd, "model")) {
		enum x52_type type = x52_gettype(hdl);
		switch (type) {
		case DEV_X52:
			printf("Model: X52 Flight Control System\n");
			break;
		case DEV_X52PRO:
			printf("Model: X52 Pro Flight Control System\n");
			break;
		case DEV_YOKE:
			printf("Model: Pro Flight Yoke\n");
			break;
		}
	} else if (!strcmp(cmd, "custom")) {
		if (argc < 4) {
			printf("usage: %s custom <idx> <value>\n", prog);
			return 0;
		}
		int index = strtol(argv[2], NULL, 16);
		int value = strtol(argv[3], NULL, 16);
		x52_custom(hdl, index, value);
	} else {
		printf("unknown command %s\n", cmd);
	}

	x52_close(hdl);
	return 0;
}
